src/Microsoft.Crank.AzureDevOpsWorker/Program.cs

```diff
@@ -344,5 +358,85 @@ private static Task ErrorHandler(ProcessErrorEventArgs args)
         }
 
         private static string LogNow => $"[{DateTime.Now.ToString("hh:mm:ss.fff")}]";
+
+        internal static void MaterializeFiles(JobPayload jobPayload, string workingDirectory)
+        {
+            if (jobPayload?.Files == null || jobPayload.Files.Count == 0)
+            {
+                return;
+            }
+
+            var baseDir = Path.GetFullPath(string.IsNullOrEmpty(workingDirectory)
+                ? Directory.GetCurrentDirectory()
+                : workingDirectory);
+
+            // Ensure base working directory exists
+            Directory.CreateDirectory(baseDir);
+
+            foreach (var file in jobPayload.Files)
+            {
+                var relativePath = file.Key?.Replace('\\', '/');
+                if (string.IsNullOrWhiteSpace(relativePath))
+                {
+                    Console.WriteLine($"{LogNow} Skipping empty file path entry.");
+                    continue;
+                }
+
+                var targetPath = Path.GetFullPath(Path.Combine(baseDir, relativePath));
+
+                // Ensure targetPath stays within baseDir to prevent path traversal
+                var normalizedBaseDir = Path.GetFullPath(baseDir).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
+                var normalizedTarget = Path.GetFullPath(targetPath).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
+
+                // Allow baseDir itself (e.g., "./") and any child path
+                var baseWithSep = normalizedBaseDir + Path.DirectorySeparatorChar;
+                if (!(normalizedTarget.Equals(normalizedBaseDir, StringComparison.OrdinalIgnoreCase) ||
+                      normalizedTarget.StartsWith(baseWithSep, StringComparison.OrdinalIgnoreCase)))
+                {
+                    Console.WriteLine($"{LogNow} Skipping unsafe path: {relativePath}");
+                    continue;
+                }
+
+                var parentDir = Path.GetDirectoryName(targetPath);
+                if (!string.IsNullOrEmpty(parentDir))
+                {
+                    Directory.CreateDirectory(parentDir);
+                }
+
+                try
+                {
+                    var fileContent = Convert.FromBase64String(file.Value ?? string.Empty);
+                    File.WriteAllBytes(targetPath, fileContent);
+                }
+                catch (FormatException)
+                {
+                    Console.WriteLine($"{LogNow} Invalid base64 content for file: {relativePath}. Skipping. Content: {file.Value} \n\n");
+                }
+            }
+        }
+
+        // Create a unique temp working directory
+        private static string CreateTempWorkingDirectory()
+        {
+            var dir = Path.Combine(Path.GetTempPath(), "crank-" + Guid.NewGuid().ToString("N"));
```

Comment:
Prefer `Directory.Createtempsubdirectory("crank-
)` because it will set the correct permissions so only the current account can access it. The tmp folder on linux is visible from everyone.

Reply:
Why resolved?

Reply:
Removed the creation of the temp directory and instead create the crank-* directory from the running directory of the devops worker.

Reply:
I meant it literally, `Directory.CreateTempSubdirectory`

Reply:
Ah, that makes sense. Do you have a preference on using the System temp with `Directory.CreateTempSubdirectory` or managing our own temp directories?